File: platforms/android/runtime/core/src/main/java/io/nativeblocks/runtime/frame/ActionTree.kt
@file:Suppress("DEPRECATION")

package io.nativeblocks.runtime.frame

import io.nativeblocks.runtime.api.provider.action.ActionContext
import io.nativeblocks.runtime.api.provider.action.INativeAction
import io.nativeblocks.runtime.api.provider.action.NativeActionProviderRegistry
import io.nativeblocks.runtime.api.provider.action.defaults.NativeScriptAction
import io.nativeblocks.runtime.api.provider.action.defaults.internalFallbackAction
import io.nativeblocks.runtime.api.provider.model.NativeActionModel
import io.nativeblocks.runtime.api.provider.model.NativeActionTriggerModel
import io.nativeblocks.runtime.api.provider.model.NativeVariableModel
import io.nativeblocks.runtime.api.provider.model.resolveIn
import io.nativeblocks.runtime.ffi.ActionLogEvent
import kotlinx.coroutines.CoroutineScope

internal class ActionTree(
    private val instanceName: String,
    private val coroutineScope: CoroutineScope,
    private val onFindVariable: (String) -> NativeVariableModel?,
    private val onVariableChange: (NativeVariableModel) -> Unit,
    private val onLog: (ActionLogEvent) -> Unit,
) {

    private val nativeActionProvider = NativeActionProviderRegistry.getOrCreate(instanceName)

    fun handle(index: Int, action: NativeActionModel?, performedEventType: String, scope: Any? = null) {
        if (action == null || action.event != performedEventType) {
            onLog(ActionLogEvent.EventIgnored(performedEventType))
            return
        }

        onLog(ActionLogEvent.EventTriggered(performedEventType, action.key))

        action.triggers
            .filter { it.parentId.isEmpty() }
            .forEach { trigger ->
                handleTrigger(
                    action = action,
                    index = index,
                    scope = scope,
                    trigger = trigger.copy(subTriggers = action.triggers.filter { it.parentId == trigger.id }),
                    onFind = { keyType ->
                        nativeActionProvider.getProvidedActions()[keyType]
                    },
                    onTriggerFallBack = { keyType, name ->
                        nativeActionProvider.getFallbackAction()?.invoke(keyType, name)
                            ?: internalFallbackAction(instanceName = instanceName, name = name)
                    }
                )
            }
    }

    private fun handleTrigger(
        action: NativeActionModel,
        index: Int,
        scope: Any?,
        trigger: NativeActionTriggerModel,
        onFind: (String) -> INativeAction?,
        onTriggerFallBack: (keyType: String, name: String) -> Unit,
    ) {
        val nativeAction = if (trigger.keyType == NativeScriptAction.KEY_TYPE) {
            NativeScriptAction()
        } else {
            onFind(trigger.keyType)
        }

        if (nativeAction == null) {
            onLog(ActionLogEvent.TriggerFallback(trigger.keyType, trigger.name))
            onTriggerFallBack(trigger.keyType, trigger.name)
            return
        }

        onLog(ActionLogEvent.TriggerExecuted(trigger.name, trigger.keyType, trigger.event))

        val actionContext = ActionContext(
            instanceName = instanceName,
            listItemIndex = index,
            coroutineScope = coroutineScope,
            trigger = trigger,
            onFindVariable = onFindVariable,
            onUpdateVariable = { variable ->
                variable?.let { onVariableChange(it) }
            },
            onHandleEvent = { event ->
                advanceSubTriggers(action, index, scope, trigger, event, onFind, onTriggerFallBack)
            },
            resolveTemplate = { value ->
                resolveIn(scope, value)
            },
        )
        nativeAction.handle(actionContext)
    }

    private fun advanceSubTriggers(
        action: NativeActionModel,
        index: Int,
        scope: Any?,
        trigger: NativeActionTriggerModel,
        event: String,
        onFind: (String) -> INativeAction?,
        onTriggerFallBack: (keyType: String, name: String) -> Unit,
    ) {
        action.triggers
            .filter { it.parentId == trigger.id && it.event == event }
            .forEach { handleTrigger(action, index, scope, it, onFind, onTriggerFallBack) }
    }
}

File: platforms/ios/runtime/Sources/NativeblocksRuntime/frame/ActionTree.swift
import Foundation
import NativeblocksRuntimeFFI

internal final class ActionTree {

    private let instanceName: String
    private let onFindVariable: (String) -> NativeVariableModel?
    private let onVariableChange: (NativeVariableModel) -> Void
    private let onLog: (ActionLogEvent) -> Void

    init(
        instanceName: String,
        onFindVariable: @escaping (String) -> NativeVariableModel?,
        onVariableChange: @escaping (NativeVariableModel) -> Void,
        onLog: @escaping (ActionLogEvent) -> Void
    ) {
        self.instanceName = instanceName
        self.onFindVariable = onFindVariable
        self.onVariableChange = onVariableChange
        self.onLog = onLog
    }

    private var nativeActionProvider: NativeActionProvider {
        return NativeActionProviderRegistry.getOrCreate(instanceName)
    }

    func handle(index: Int, action: NativeActionModel?, performedEventType: String, scope: Any? = nil) {
        guard let action, action.event == performedEventType else {
            onLog(.eventIgnored(event: performedEventType))
            return
        }

        onLog(.eventTriggered(event: performedEventType, actionKey: action.key))

        for trigger in action.triggers where trigger.parentId.isEmpty {
            var rootTrigger = trigger
            rootTrigger.subTriggers = action.triggers.filter { $0.parentId == trigger.id }
            handleTrigger(
                action: action,
                index: index,
                scope: scope,
                trigger: rootTrigger,
                onFind: { [weak self] keyType in
                    self?.nativeActionProvider.getProvidedActions()[keyType]
                },
                onTriggerFallBack: { [weak self] keyType, name in
                    guard let self else { return }
                    if let fallback = self.nativeActionProvider.getFallbackAction() {
                        fallback(keyType, name)
                    } else {
                        internalFallbackAction(instanceName: self.instanceName, name: name)
                    }
                }
            )
        }
    }

    private func handleTrigger(
        action: NativeActionModel,
        index: Int,
        scope: Any?,
        trigger: NativeActionTriggerModel,
        onFind: @escaping (String) -> (any INativeAction)?,
        onTriggerFallBack: @escaping (String, String) -> Void
    ) {
        let nativeAction: (any INativeAction)? =
            trigger.keyType == NativeScriptAction.KEY_TYPE ? NativeScriptAction() : onFind(trigger.keyType)

        guard let nativeAction else {
            onLog(.triggerFallback(keyType: trigger.keyType, name: trigger.name))
            onTriggerFallBack(trigger.keyType, trigger.name)
            return
        }

        onLog(
            .triggerExecuted(
                name: trigger.name,
                keyType: trigger.keyType,
                event: trigger.event
            )
        )

        let actionContext = ActionContext(
            instanceName: instanceName,
            listItemIndex: index,
            onFindVariable: onFindVariable,
            onUpdateVariable: { [weak self] variable in
                guard let variable else { return }
                self?.onVariableChange(variable)
            },
            trigger: trigger,
            onHandleEvent: { [weak self] event in
                self?.advanceSubTriggers(
                    action: action, index: index, scope: scope, trigger: trigger, event: event,
                    onFind: onFind, onTriggerFallBack: onTriggerFallBack
                )
            },
            resolveTemplate: { resolveIn(scope, $0) }
        )
        nativeAction.handle(actionContext: actionContext)
    }

    private func advanceSubTriggers(
        action: NativeActionModel,
        index: Int,
        scope: Any?,
        trigger: NativeActionTriggerModel,
        event: String,
        onFind: @escaping (String) -> (any INativeAction)?,
        onTriggerFallBack: @escaping (String, String) -> Void
    ) {
        for subTrigger in action.triggers where subTrigger.parentId == trigger.id && subTrigger.event == event {
            handleTrigger(
                action: action,
                index: index,
                scope: scope,
                trigger: subTrigger,
                onFind: onFind,
                onTriggerFallBack: onTriggerFallBack
            )
        }
    }
}
